src/lib/core/datetime.h

```diff
@@ -77,6 +77,57 @@ tnt_datetime_strftime(const struct datetime *date, char *buf, size_t len,
 void
 tnt_datetime_now(struct datetime *now);
 
+/**
+ * Local version resembling ISO C' tm structure.
+ * Includes original epoch value, and nanoseconds.
+ */
+struct tnt_tm {
```

Comment:
Almost good! But you misunderstood me a bit. You still have the cyclic dependency. datetime and tzcode still depend on each other. `tzcode` needs `struct tnt_tm` and `datetime` needs `tnt_strftime`. It isn't very much better now. You so far just renamed `datetime_tm` into `tnt_tm`.

This is how it should look.
`tzcode` is a self-sufficient library. It never touches datetime.h, datetime.c. It has `struct tnt_tm` definition in its header somewhere. I see in your patch `tzcode` does not have a header - maybe you need to add one. The library works with `stuct tnt_tm` only. It is very important. `tzcode` has **0 dependencies on datetime**.

`datetime` library **depends** on `tzcode` library. It is built on top of it. It includes tzcode's header in datetime.c. `datetime_strftime` is defined in datetime.c. It converts `struct datetime` into `struct tnt_tm` and calls `tnt_strftime` with this tm struct.

Summary:
- `datetime` has: `datetime_strftime(datetime *)`, which uses `datetime_to_tnt_tm(datetime *, tnt_tm *)`. It depends on `tzcode` library.
- `tzcode` has `struct tnt_tm` and `tnt_strftime(struct tm *)`. It does not depend on anything anyhow.

Reply:
1. I've moved `third_party/tzcode` to `src/lib/tzcode`, is we have modified `strftime.c` very heavily, and it's very different to the original IANA `tzcode/strftime.c` implementation. Also we foresee to add `strptime.c` from FreeBSD and all the necessary pieces for `tzindex` support from Olson (also to be heavily modified). 
2. I've created `tzcode/tzcode.h` where I've put definitions of `struct tnt_tm` and `tnt_strftime()` (more to come in next commits).

<details>
  <summary>`tzcode.h` introduction</summary>

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1bbfba25a..8037c30a7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -579,14 +579,6 @@ include(BuildCDT)
 libccdt_build()
 add_dependencies(build_bundled_libs cdt)
 
-#
-# portion of Olson tzcode
-#
-
-include(BuildTZ)
-libtzcode_build()
-add_dependencies(build_bundled_libs tzcode)
-
 #
 # Third-Party misc
 #
diff --git a/cmake/BuildTZ.cmake b/cmake/BuildTZ.cmake
deleted file mode 100644
index dd4733300..000000000
--- a/cmake/BuildTZ.cmake
+++ /dev/null
@@ -1,5 +0,0 @@
-macro(libtzcode_build)
-    set(LIBTZ_LIBRARIES tzcode)
-
-    add_subdirectory(${PROJECT_SOURCE_DIR}/third_party/tzcode)
-endmacro()
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 38b685b6b..168ac2a52 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -191,14 +191,13 @@ set_source_files_compile_flags(${server_sources})
 add_library(server STATIC ${server_sources})
 add_dependencies(server build_bundled_libs)
 target_link_libraries(server core coll http_parser bit uri uuid swim swim_udp
-                      swim_ev crypto mpstream crc32)
+                      swim_ev crypto mpstream crc32 tzcode)
 
 # Rule of thumb: if exporting a symbol from a static library, list the
 # library here.
 set (reexport_libraries server core misc bitset csv swim swim_udp swim_ev
-     shutdown ${LUAJIT_LIBRARIES} ${MSGPUCK_LIBRARIES} ${ICU_LIBRARIES}
-     ${CURL_LIBRARIES} ${XXHASH_LIBRARIES} ${LIBCDT_LIBRARIES}
-     ${LIBTZ_LIBRARIES})
+     shutdown tzcode ${LUAJIT_LIBRARIES} ${MSGPUCK_LIBRARIES} ${ICU_LIBRARIES}
+     ${CURL_LIBRARIES} ${XXHASH_LIBRARIES} ${LIBCDT_LIBRARIES})
 
 set (common_libraries
     ${reexport_libraries}
diff --git a/src/lib/CMakeLists.txt b/src/lib/CMakeLists.txt
index 06f935f3e..81ccc585d 100644
--- a/src/lib/CMakeLists.txt
+++ b/src/lib/CMakeLists.txt
@@ -16,6 +16,7 @@ add_subdirectory(mpstream)
 add_subdirectory(vclock)
 add_subdirectory(raft)
 add_subdirectory(fakesys)
+add_subdirectory(tzcode)
 if(ENABLE_BUNDLED_MSGPUCK)
     add_subdirectory(msgpuck EXCLUDE_FROM_ALL)
 endif()
diff --git a/src/lib/core/CMakeLists.txt b/src/lib/core/CMakeLists.txt
index 8bc776b82..948659ab4 100644
--- a/src/lib/core/CMakeLists.txt
+++ b/src/lib/core/CMakeLists.txt
@@ -42,10 +42,10 @@ endif()
 
 add_library(core STATIC ${core_sources})
 
-target_link_libraries(core salad small uri decNumber bit ${LIBEV_LIBRARIES}
+target_link_libraries(core salad small uri decNumber bit tzcode
+                      ${LIBEV_LIBRARIES}
                       ${LIBEIO_LIBRARIES} ${LIBCORO_LIBRARIES}
-                      ${MSGPUCK_LIBRARIES} ${ICU_LIBRARIES}
-                      ${LIBTZ_LIBRARIES})
+                      ${MSGPUCK_LIBRARIES} ${ICU_LIBRARIES})
 
 if (ENABLE_BACKTRACE AND NOT TARGET_OS_DARWIN)
     target_link_libraries(core gcc_s ${UNWIND_LIBRARIES})diff --git a/src/lib/core/datetime.h b/src/lib/core/datetime.h
index a2fcaf57f..13f7d3533 100644
--- a/src/lib/core/datetime.h
+++ b/src/lib/core/datetime.h
@@ -8,6 +8,8 @@
 #include <stdint.h>
 #include <sys/types.h>
 
+#include "tzcode/tzcode.h"
+
 #if defined(__cplusplus)
 extern "C" {
 #endif /* defined(__cplusplus) */
@@ -77,57 +79,12 @@ tnt_datetime_strftime(const struct datetime *date, char *buf, size_t len,
 void
 tnt_datetime_now(struct datetime *now);
 
-/**
- * Local version resembling ISO C' tm structure.
- * Includes original epoch value, and nanoseconds.
- */
-struct tnt_tm {
-	/** Seconds. [0-60] (1 leap second) */
-	int tm_sec;
-	/** Minutes. [0-59] */
-	int tm_min;
-	/** Hours. [0-23] */
-	int tm_hour;
-	/** Day. [1-31] */
-	int tm_mday;
-	/** Month. [0-11] */
-	int tm_mon;
-	/** Year - 1900. */
-	int tm_year;
-	/** Day of week. [0-6] */
-	int tm_wday;
-	/** Days in year.[0-365] */
-	int tm_yday;
-	/** DST. [-1/0/1] */
-	int tm_isdst;
-
-	/** Seconds east of UTC. */
-	long int tm_gmtoff;
-	/** Timezone abbreviation. */
-	const char *tm_zone;
-	/** Seconds since Epoch */
-	int64_t tm_epoch;
-	/** nanoseconds */
-	int tm_nsec;
-};
-
 /**
  * Convert @sa datetime to @sa dt_tm
  */
 void
 datetime_to_tm(const struct datetime *date, struct tnt_tm *tm);
 
-/**
- * tnt_strftime is Tarantool version of a POSIX' strftime()
- * which has been extended with %f (fractions of second)
- * flag support. In all other aspect it's behaving exactly
- * like standard strftime.
- * @sa strftime()
- */
-size_t
-tnt_strftime(char *s, size_t maxsize, const char *format,
-	     const struct tnt_tm *tm);
-
 #if defined(__cplusplus)
 } /* extern "C" */
 #endif /* defined(__cplusplus) */
diff --git a/src/lib/tzcode/strftime.c b/src/lib/tzcode/strftime.c
index 6fd8d1001..c006ffe25 100644
--- a/src/lib/tzcode/strftime.c
+++ b/src/lib/tzcode/strftime.c
@@ -33,7 +33,6 @@
 ** This is ANSIish only when "multibyte character == plain character".
 */
 
-#include "datetime.h"
 #include "private.h"
 #include "trivia/util.h"
 #include "tzcode.h"
diff --git a/third_party/tzcode/CMakeLists.txt b/src/lib/tzcode/CMakeLists.txt
similarity index 69%
rename from third_party/tzcode/CMakeLists.txt
rename to src/lib/tzcode/CMakeLists.txt
index f76721231..0ed4ffba6 100644
--- a/third_party/tzcode/CMakeLists.txt
+++ b/src/lib/tzcode/CMakeLists.txt
@@ -2,4 +2,4 @@ set(tzcode_sources strftime.c)
 
 add_library(tzcode STATIC ${tzcode_sources})
 
-target_link_libraries(tzcode core)
+target_link_libraries(tzcode)
diff --git a/third_party/tzcode/private.h b/src/lib/tzcode/private.h
similarity index 100%
rename from third_party/tzcode/private.h
rename to src/lib/tzcode/private.h
diff --git a/third_party/tzcode/strftime.c b/src/lib/tzcode/strftime.c
similarity index 94%
rename from third_party/tzcode/strftime.c
rename to src/lib/tzcode/strftime.c
index 438882560..6fd8d1001 100644
--- a/third_party/tzcode/strftime.c
+++ b/src/lib/tzcode/strftime.c
@@ -33,9 +33,10 @@
 ** This is ANSIish only when "multibyte character == plain character".
 */
 
-#include "trivia/util.h"
-#include "datetime.h"
 #include "private.h"
+#include "trivia/util.h"
+#include "tzcode.h"
 
 #include <assert.h>
 #include <locale.h>
@@ -108,36 +109,36 @@ enum warn {
 	IN_ALL
 };
 
-static size_t
-_add(char *buf, size_t size, const char *str);
-static size_t
-_conv(char *buf, size_t size, const char *format, int n);
-static size_t
-_yconv(char *buf, size_t size, int a, int b, bool convert_top, bool convert_yy);
-static size_t
-_fmt(char *buf, size_t size, const char *format, const struct tnt_tm *t,
+static ssize_t
+_add(char *buf, ssize_t size, const char *str);
+static ssize_t
+_conv(char *buf, ssize_t size, const char *format, int n);
+static ssize_t
+_yconv(char *buf, ssize_t size, int a, int b, bool convert_top, bool convert_yy);
+static ssize_t
+_fmt(char *buf, ssize_t size, const char *format, const struct tnt_tm *t,
      enum warn *warnp);
 
-
 size_t
-tnt_strftime(char *s, size_t maxsize, const char *format,
+tnt_strftime(char *s, ssize_t maxsize, const char *format,
 	     const struct tnt_tm *t)
 {
 	tzset();
 	enum warn warn = IN_NONE;
 
-	size_t total = _fmt(s, maxsize, format, t, &warn);
-	if (total < maxsize)
+	ssize_t total = _fmt(s, maxsize, format, t, &warn);
+	assert(total >= 0);
+	if (total >= 0 && total < maxsize)
 		s[total] = '\0';
 
-	return total;
+	return (size_t)total;
 }
 
-static size_t
-_fmt(char *buf, size_t size, const char *format, const struct tnt_tm *t,
+static ssize_t
+_fmt(char *buf, ssize_t size, const char *format, const struct tnt_tm *t,
      enum warn *warnp)
 {
-	size_t total = 0;
+	ssize_t total = 0;
 
 	for (; *format; ++format) {
 		if (*format == '%') {
@@ -583,18 +584,18 @@ _fmt(char *buf, size_t size, const char *format, const struct tnt_tm *t,
 	return total;
 }
 
-static size_t
-_conv(char *buf, size_t size, const char *format, int n)
+static ssize_t
+_conv(char *buf, ssize_t size, const char *format, int n)
 {
-	size_t total = 0;
+	ssize_t total = 0;
 	SNPRINT(total, snprintf, buf, size, format, n);
 	return total;
 }
 
-static size_t
-_add(char *buf, size_t size, const char *str)
+static ssize_t
+_add(char *buf, ssize_t size, const char *str)
 {
-	size_t total = 0;
+	ssize_t total = 0;
 	SNPRINT(total, snprintf, buf, size, "%s", str);
 	return total;
 }
@@ -607,12 +608,12 @@ _add(char *buf, size_t size, const char *str)
 ** with more only if necessary.
 */
 
-static size_t
-_yconv(char *buf, size_t size, int a, int b, bool convert_top, bool convert_yy)
+static ssize_t
+_yconv(char *buf, ssize_t size, int a, int b, bool convert_top, bool convert_yy)
 {
 	register int lead;
 	register int trail;
-	size_t total = 0;
+	ssize_t total = 0;
 diff --git a/src/lib/tzcode/tzcode.h b/src/lib/tzcode/tzcode.h
new file mode 100644
index 000000000..26198d524
--- /dev/null
+++ b/src/lib/tzcode/tzcode.h
@@ -0,0 +1,62 @@
+#pragma once
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2021, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#include <stdint.h>
+#include <sys/types.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+/**
+ * Local version resembling ISO C' tm structure.
+ * Includes original epoch value, and nanoseconds.
+ */
+struct tnt_tm {
+       /** Seconds. [0-60] (1 leap second) */
+       int tm_sec;
+       /** Minutes. [0-59] */
+       int tm_min;
+       /** Hours. [0-23] */
+       int tm_hour;
+       /** Day. [1-31] */
+       int tm_mday;
+       /** Month. [0-11] */
+       int tm_mon;
+       /** Year - 1900. */
+       int tm_year;
+       /** Day of week. [0-6] */
+       int tm_wday;
+       /** Days in year.[0-365] */
+       int tm_yday;
+       /** DST. [-1/0/1] */
+       int tm_isdst;
+
+       /** Seconds east of UTC. */
+       long int tm_gmtoff;
+       /** Timezone abbreviation. */
+       const char *tm_zone;
+       /** Seconds since Epoch */
+       int64_t tm_epoch;
+       /** nanoseconds */
+       int tm_nsec;
+};
+
+/**
+ * tnt_strftime is Tarantool version of a POSIX' strftime()
+ * which has been extended with %f (fractions of second)
+ * flag support. In all other aspect it's behaving exactly
+ * like standard strftime.
+ * @sa strftime()
+ */
+size_t
+tnt_strftime(char *s, ssize_t maxsize, const char *format,
+            const struct tnt_tm *tm);
+
+#if defined(__cplusplus)
+} /* extern "C" */
+#endif /* defined(__cplusplus) */
```
</details>